### PR TITLE
create isStructure function

### PR DESCRIPTION
**Pull request by lublak**
It would be practical to test only on one structure to pretest methods that say "method is only guaranteed to work on anonymous structures".

### CODE REVIEW

**Comment by lublak** [comment]
Since flash is not tested via CI, is flash still supported at all and requires implementation?

File: std/lua/_std/Reflect.hx [hunk 1]
```diff
@@ -121,6 +121,15 @@ import lua.Boot;
 				|| (t == "function" && (lua.Boot.isClass(v) || lua.Boot.isEnum(v)) != null);
 		}
 
+	public static function isStructure(v:Dynamic):Bool untyped {
+		if (v == null) return false;
+		if(Lua.type(v) != "table" || v.__enum__ != null || lua.Boot.isClass(v) || lua.Boot.isEnum(v)) return false;
+		var mt = Lua.getmetatable(v);
+		if(mt == null) return true;
+		if(mt.__index != null) return false;
+		return !Reflect.hasField(mt, '__index');
```
Comment (lublak) [inline std/lua/_std/Reflect.hx:130]: this is needed because `Type.createEmptyInstance(ReallyEmptyClass)` __index is null

**Comment by Simn** [comment]
Hmm, I don't know if we want the concept of "structure" to exist at run-time at all...

**Comment by lublak** [comment]
@Simn but currently there is no simple option to check if a dynamic value is a anonym structur.
It would be usefull to test it before a Reflect.fields/Reflect.setField etc.

**Comment by RealyUniqueName** [comment]
There's `Type.typeof(v) == TObject`. 
In general there's no way to be 100% sure an object is a Haxe anonymous structure. It's possible only on some targets. 
If anything this should be a method in `<target>.Lib`

**Comment by kevinresol** [comment]
> In general there's no way to be 100% sure an object is a Haxe anonymous structure. It's possible only on some targets.

Example: `Reflect.isStructure(new js.lib.Map())`

**Comment by lublak** [comment]
first of all you use `Type.typeof(v)` which creates on some platforms more overhead.
Also it generates on some targets also for some classes "TObject".
So it is more usefull to check if something can be Relfection.setField. Of course you can't check if its native or a haxe structur. But my method should return false on classes, functions, etc. 
Maybe the name is badly chosen. But currently there is no easy way to check if it is an "anonymous structure" or a class.
Js is currenlty the only target which is very hard to implement it but it is also possible to check if it is an object or a class object.
And sorry if my english is not perfect. So if something was not properly understandable please ask again.
`Reflect.isFieldable` `Reflect.isModifiable` `Reflect.isLikeStruct`

**Comment by RealyUniqueName** [comment]
I'd say in such cases one should use `try...catch`
```haxe
try {
  Reflect.setField(o, 'field', value);
} catch(_) {
  trace('Reflect.setField does not work with object $o');
}
```

**Comment by Aurel300** [comment]
> I'd say in such cases one should use `try...catch`

Requiring an exception to do something you can do on most targets in a cheaper way seems like a bad design. I don't see a huge problem with this API?

**Comment by lublak** [comment]
The function is currently fully functional. But I am also okay with it if it is not wanted.
Basically I wanted to provide a way to check if it is a haxe generated structure or not.
If it is not wanted I am fine with it and will adopt it in my library (except unfortunately the eval function).

**Comment by Simn** [comment]
> Requiring an exception to do something you can do on most targets in a cheaper way seems like a bad design. I don't see a huge problem with this API?

Maybe, but the standard for new API additions is a bit higher than "there's not a huge problem". Especially if it imposes additional run-time constraints on targets. 

If the goal here is to answer "can I use setField on it" then the implementation of `setField` should be changed to `if (canIUseSetFieldOn(it)) setTheField()`. Otherwise there's a potential disconnect. Of course you then have the problem that actually calling `setField` would run the check again, and at that point it gets rather messy and complicated.

Overall I'm not in favor of this change until somebody can present a convincing use-case _and_ a clean implementation. I don't really mind that we're starting with the implementation here because that's quite pragmatic, but we still have to make sure that we actually want this.

**Comment by lublak** [comment]
Possibly the other way round?
isClassInstance()?
So you can check: `isObject && !isFunction && !isEnumValue && !isClass && !isEnum`
(Yes it's generate more overhead than isStructure, but maybe it fits more into the framework of Haxe.)
Also it has some issues because it's only checks for haxe classes? Could be a problem.

**Comment by lublak** [comment]
If the vote is completely against it, the next question follows:
The parts of the code that can 100% determine whether it is a structure or not. Should I then transfer them to the respective Lib.
(All except js and Lua)
Basically, the reflection libraries do not always work 100% on the targets when working with native types. Especially in the js area, many things could occur with the type and reflect functions if a developer would set the respective fields Manuel. Therefore, isStructur is actually also possible there if the developer does not mess around in the native code.

**Comment by lublak** [comment]
> 
> 
> > In general there's no way to be 100% sure an object is a Haxe anonymous structure. It's possible only on some targets.
> 
> Example: `Reflect.isStructure(new js.lib.Map())`

Currently returns false because `js.lib.Map` is a class.
Which is actually correct, isn't it?

**Comment by lublak** [comment]
Here is an example of what it can be used for:
```haxe
public static function isType(v:Dynamic, t:Dynamic) {
    return switch Type.typeof(t) {
      case TNull: v == null;
      case TInt: v == t;
      case TFloat: v == t;
      case TBool: v == t;
      case TObject:
        Reflect.isStruct(t) ? someMethodToCompareStructs(v, t) : Std.isOfType(v, t);
      case TFunction: Reflect.compareMethods(v, t);
      default: Std.isOfType(v, t);
    }
  }
```

**Comment by RealyUniqueName** [comment]
I don't like this change at all.
It's not only about API. It adds new implications and constraints.

**Comment by lublak** [comment]
What would be an alternative idea to check whether an object is a structure?

**Comment by RealyUniqueName** [comment]
I don't know any other ways to check that.

**Comment by lublak** [comment]
@RealyUniqueName would be the alternative:
- isClass
- isEnum
be a better alternative.
Basically I want to know that if it is an object but not a function, class or enum.

**Comment by Simn** [comment]
I'm making a decision here to not support this. Targets shouldn't be required to track the origin of a run-time class, and stuff like this varies from target to target anyway and might even be influenced by optimization. I suggest that you instead start looking at the actual problem you want to solve, and then present that as an issue so we can discuss possible solutions. Sorry if that makes you feel like you wasted your time on this implementation.

**Comment by lublak** [comment]
@Simn all fine :) 
And also have understanding for it.

**Pull request closed without merging**